pyproject.toml

```diff
@@ -1,27 +1,44 @@
 [tool.poetry]
 name = "tsdfileapi"
-version = "2.10.4"
-description = ""
-authors = ["Your Name <you@example.com>"]
+version = "0.0.0"
+description = "A REST API for upload and download of files and JSON data."
+authors = [
+    "Leon du Toit <l.c.d.toit@usit.uio.no>",
+    "Eirik Haatveit <haatveit@uio.no>",
+    "Milen Kouylekov <milen.kouylekov@usit.uio.no>",
+    "Armen Michaeli <armen.michaeli@usit.uio.no>",
+]
+
+[tool.poetry.scripts]
+tsd-file-api = "tsdfileapi.api:main"
 
 [tool.poetry.dependencies]
-python = "^3.9"
-jwcrypto = "^1.3.1"
-tornado = "^6.2"
-psycopg2-binary = "^2.9.3"
-libnacl = "^1.8.0"
-pysquril = {url = "https://github.com/unioslo/pysquril/archive/refs/tags/v0.1.4.tar.gz"}
-tsd-api-client = "^3.5.2"
-SQLAlchemy = "^1.4.40"
-pretty-bad-protocol = "^3.1.1"
-click = "^8.1.3"
-termcolor = "^1.1.0"
-python-magic = "^0.4.27"
-pika = "^1.3.0"
+python = "^3.7"
+jwcrypto = ">=1.4"
```

Comment:
I think we should have pinned versions, so we can do controlled upgrades - ISO, all about control

Reply:
We should definitely be pinning versions, but my intention was that this would be handled by the `poetry.lock` file (unignored in c9a0bb48cbadf956760c55385750caa125b14d21 and then committed in 6807222db8a5b5a83a5b7d100394e51565f6ff9b).

If a `poetry.lock` file is present in a project, it will be used to [install the exact versions listed there](https://python-poetry.org/docs/basic-usage/#installing-with-poetrylock) when running `poetry install`. Likewise, `poetry export -f requirements.txt` with a lock file present will dump the locked dependencies rather than perform any resolution itself.

`poetry update` or `poetry lock` are used to resolve dependencies to newer versions within constraints defined in `pyproject.toml`, and write this out to the lock file, but no dependencies will be upgraded unless explicitly asked to do so. As the `poetry.lock` file is committed to the repository in this PR, any dependency version updates will need to be explicitly requested and PR'd by a developer to apply.

It was actually my intention to submit a follow-up PR next to loosen all the dependency version constraints, and then addressing #131 by updating the lock file and verifying that all tests still pass.

I really believe this should give us the type of control we need, but please let me hear your thoughts.

Reply:
Assume the following constraints apply:
```
dependency1 = ">=0.1"
dependency2 = "^0"
dependency3 = ">=1.2"
```

Now suppose `dependency1` has published versions 0.2 and 0.3, and we have currently 0.1. If I want to update `dependency1` from 0.1 to 0.2, _without updating any other dependencies_, how do I do that with the proposed regime?

Reply:
In this case we would need to change the constraints of `dependency1` as `>=0.1` would give us version 0.3 rather than 0.2. We'd then need to alter the dependency specification in `pyproject.toml` to the following:

```
dependency1 = ">=0.1,<0.3"
```

And afterwards run `poetry update dependency1` to only update this dependency (and its dependencies if required).

Reply:
Wouldn't that also update the other dependencies if new versions are available, given their constraints?

Reply:
If you run `poetry update <package>` it will only update the one package that you passed as a parameter, according to its constraints.

If we don't *need* highly specific dependencies for some reason, then having loose constraints in `pyproject.toml` makes it easier to keep the dependencies up to date while at the same time maintaining a consistent, reproducible shared environment via the lock file.

Updating dependencies would then be a matter of:

1. running `poetry update`
2. making sure all tests still clear, and if not
  a. change dependency constraints
  b. patch our usage of the dependency
3. commit to repo and submit PR

When I say using looser constraints is *easier* I'm not saying that the alternative would be *hard*, just that it's more work if we have to manually look up each package to find out what they've been up to recently. To me that just seems to make us less likely to keep our dependencies up to date.

In the example of the `dependency1` above, I assume that there's a technical reason for why we would want version 0.2 specifically and not 0.3 even though it is the latest. In that case it very much makes sense to then introduce a constraint that keeps us on the 0.2 release series until a usable new version is published.

As you have probably guessed by now I'm not a huge fan of hard pinned versions, as trying to combine packages that have specific version requirements set will inevitably lead to resolution failure if they share any dependencies between them. I've had this issue many times in the past when importing packages that opted for this kind of dependency specification strategy, forcing me to rewrite half of my dependency list to be able to resolve after. Presumably the file API will not be imported elsewhere though, so this will not really give us issue unless tsd-api-client or pysquril (or any of the externally developed packages) also start using highly specific dependencies.

Unlike our previous `requirements.txt`, `poetry.lock` defines a fully reproducible environment with our direct dependencies as well as dependencies of dependencies defined, similar to `pip freeze`. I'd argue that we are improving control and reproducibility either way we go with this.

Reply:
As dicussed in DMs, I think it is fine to use the approach you've outlined, but we should make sure we have the proper routines to support proper dependency management - i.e., if we aim to update a specific dependency, and the result of running `poetry update` is that we suddenly update another, then we must read the changelog on all relevant updates before accepting the changes.